nexus/src/app/sagas/snapshot_create.rs

```diff
@@ -222,6 +280,187 @@ impl NexusSaga for SagaSnapshotCreate {
 
 // snapshot create saga: action implementations
 
+async fn ssc_save_disk_gen(
+    sagactx: NexusActionContext,
+) -> Result<Generation, ActionError> {
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params::<Params>()?;
+
+    let opctx = OpContext::for_saga_action(&sagactx, &params.serialized_authn);
+
+    let (.., disk) = LookupPath::new(&opctx, &osagactx.datastore())
+        .project_id(params.project_id)
+        .disk_name(&params.create_params.disk.clone().into())
+        .fetch()
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    Ok(disk.runtime_state.gen)
+}
+
+async fn ssc_alloc_regions(
+    sagactx: NexusActionContext,
+) -> Result<Vec<(db::model::Dataset, db::model::Region)>, ActionError> {
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params::<Params>()?;
+    let destination_volume_id =
+        sagactx.lookup::<Uuid>("destination_volume_id")?;
+
+    // Ensure the destination volume is backed by appropriate regions.
+    //
+    // This allocates regions in the database, but the disk state is still
+    // "creating" - the respective Crucible Agents must be instructed to
+    // allocate the necessary regions before we can mark the disk as "ready to
+    // be used".
+    //
+    // TODO: Depending on the result of
+    // https://github.com/oxidecomputer/omicron/issues/613 , we
+    // should consider using a paginated API to access regions, rather than
+    // returning all of them at once.
+    let opctx = OpContext::for_saga_action(&sagactx, &params.serialized_authn);
+
+    let (.., disk) = LookupPath::new(&opctx, &osagactx.datastore())
+        .project_id(params.project_id)
+        .disk_name(&params.create_params.disk.clone().into())
+        .fetch()
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    if disk.runtime_state.gen != sagactx.lookup::<Generation>("disk_gen")? {
```

Comment:
I'm not sure this is worth checking here...if it has changed at this point, we'll wind up catching it later and unwinding everything anyway, right?  Relatedly, it can immediately change after we do the lookup, so checking here can't be necessary to proceed.

Reply:
right, fixed in 4ebe2979